raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -565,20 +583,62 @@ private FetchQuorumRecordsResponseData buildEmptyFetchQuorumRecordsResponse(
      * - {@link Errors#INVALID_REQUEST} if the request epoch is larger than the leader's current epoch
      *     or if either the fetch offset or the last fetched epoch is invalid
      */
-    private FetchQuorumRecordsResponseData handleFetchQuorumRecordsRequest(
+    private CompletableFuture<FetchQuorumRecordsResponseData> handleFetchQuorumRecordsRequest(
         RaftRequest.Inbound requestMetadata
     ) throws IOException {
         FetchQuorumRecordsRequestData request = (FetchQuorumRecordsRequestData) requestMetadata.data;
+        FetchQuorumRecordsResponseData response = tryCompleteFetchQuorumRecordsRequest(request);
+        if (response.errorCode() != Errors.NONE.code()
+            || response.records().hasRemaining()
```

Comment:
Could this ever be true? We `serializeRecords(records)` and even if that is empty, we would not have remaining bytes right?

Reply:
We do have test cases which hit this 😉. I'm not sure I understand why you think `hasRemaining` is incorrect.

Reply:
Sorry my bad.. I was thinking totally in the opposite way.